File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.py
import sys
sys.setrecursionlimit(10**6)

class Node:
    def __init__(self, num, x, y):
        self.num = num
        self.x = x
        self.y = y
        self.left = None
        self.right = None

        
def build_tree(nodes):
    root = nodes[0]
    
    for node in nodes[1:]:
        insert(root, node)
    
    return root


def insert(parent, node):
    if(node.x < parent.x):
        if not parent.left:
            parent.left = node
        else:
            insert(parent.left, node)
    else:
        if not parent.right:
            parent.right = node
        else:
            insert(parent.right, node)

            
def get_pre_order(node, pre_order):            
    if not node:
        return
    
    pre_order.append(node.num)
    get_pre_order(node.left, pre_order)
    get_pre_order(node.right, pre_order)
    
    
def get_post_order(node, post_order):
    if not node:
        return
    
    get_post_order(node.left, post_order)
    get_post_order(node.right, post_order)
    post_order.append(node.num)
    

def solution(nodeinfo):
    
    n = len(nodeinfo)
    
    nodes = [Node(i + 1, nodeinfo[i][0], nodeinfo[i][1]) for i in range(n)]
    
    nodes.sort(key=lambda node: (-node.y, node.x))
    
    root = build_tree(nodes)
    
    pre_order = []
    post_order = []
    
    get_pre_order(root, pre_order)
    get_post_order(root, post_order)
    
    return [pre_order, post_order]

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.Arrays;

class Solution {
    
    static int[] preOrder, postOrder;
    static int order;
    
    public int[][] solution(int[][] nodeinfo) {
        
        int n = nodeinfo.length;
        Node[] nodes = new Node[n];

        for (int i = 0; i < n; i++) {
            nodes[i] = new Node(i + 1, nodeinfo[i][0], nodeinfo[i][1]);
        }

        Arrays.sort(nodes, (a, b) -> b.y - a.y);

        Node root = buildTree(nodes);

        order = 0;
        preOrder = new int[n];
        getPreOrder(root);

        order = 0;
        postOrder = new int[n];
        getPostOrder(root);

        return new int[][]{
                preOrder, postOrder
        };
    }
    
    public void getPostOrder(Node node) {
        
        if (node == null) return;

        getPostOrder(node.left);
        getPostOrder(node.right);
        postOrder[order++] = node.num;
    }

    public void getPreOrder(Node node) {

        if (node == null) return;

        preOrder[order++] = node.num;
        getPreOrder(node.left);
        getPreOrder(node.right);
    }
    
    public Node buildTree(Node[] nodes) {

        Node root = nodes[0];

        for (int i = 1; i < nodes.length; i++) {
            insert(root, nodes[i]);
        }

        return root;
    }
    
    public void insert(Node parent, Node node) {
        
        if (node.x < parent.x) {
            if (parent.left == null) {
                parent.left = node;
            } else {
                insert(parent.left, node);
            }
        } 
        else {
            if (parent.right == null) {
                parent.right = node;
            } else {
                insert(parent.right, node);
            }
        }
    }
    
    static class Node {

        int num, x, y;
        Node left, right;

        public Node(int num, int x, int y) {
            this.num = num;
            this.x = x;
            this.y = y;
        }
    }
}